File: agent-defender-js/lib/providers.js
import { FirewallOpenAI } from "./client.js";
import { loadPolicy } from "./policy.js";
import { checkToolCalls } from "./rules.js";
import { scanAndRedact } from "./pii.js";

export const OPENAI_COMPATIBLE_BASE_URLS = {
  openai: "https://api.openai.com/v1",
  groq: "https://api.groq.com/openai/v1",
  nvidia: "https://integrate.api.nvidia.com/v1",
  mistral: "https://api.mistral.ai/v1",
  together: "https://api.together.xyz/v1",
  fireworks: "https://api.fireworks.ai/inference/v1",
  perplexity: "https://api.perplexity.ai",
  deepseek: "https://api.deepseek.com",
  openrouter: "https://openrouter.ai/api/v1",
  local: "http://localhost:8000/v1",
};

export function openAICompatibleOptions(provider, opts = {}) {
  const baseURL = opts.baseURL || OPENAI_COMPATIBLE_BASE_URLS[provider];
  if (!baseURL) {
    throw new Error(`Unknown OpenAI-compatible provider: ${provider}`);
  }
  return {
    ...opts,
    baseURL,
  };
}

export async function createFirewallOpenAICompatible(provider, opts = {}) {
  const { default: OpenAI } = await import("openai");
  const { policyPath, ...clientOpts } = opts;
  const raw = new OpenAI(openAICompatibleOptions(provider, clientOpts));
  return new FirewallOpenAI(raw, { policyPath: policyPath || "policy.yaml" });
}

function stringifyArgs(args) {
  if (typeof args === "string") return args;
  if (args == null) return "{}";
  try {
    return JSON.stringify(args);
  } catch {
    return String(args);
  }
}

function read(obj, path) {
  let cur = obj;
  for (const key of path) {
    if (cur == null) return undefined;
    cur = cur[key];
  }
  return cur;
}

function makeFirewallMeta() {
  return {
    action: "allow",
    reason: null,
    rule_fired: null,
    stripped_tool_calls: [],
    blocked_calls: [],
  };
}

function enforceToolCalls(nativeCalls, policy, removeBlocked) {
  const meta = makeFirewallMeta();
  if (nativeCalls.length === 0) return meta;

  const mapped = nativeCalls.map((call) => ({
    id: call.id,
    function: {
      name: call.name,
      arguments: stringifyArgs(call.arguments),
    },
  }));

  const { findings } = checkToolCalls(mapped, policy);
  const blockedIds = new Set(
    findings.filter((f) => f.status === "block").map((f) => f.toolCallId)
  );

  if (blockedIds.size === 0) return meta;

  removeBlocked(blockedIds);
  const firstBlocked = findings.find((f) => f.status === "block");
  meta.action = "block";
  meta.stripped_tool_calls = [...blockedIds].sort();
  meta.rule_fired = "deterministic_rules";
  meta.reason = firstBlocked?.reasons?.join("; ") || "tool call blocked";
  meta.blocked_calls = findings
    .filter((f) => f.status === "block")
    .map((f) => ({
      name: f.toolName,
      reasons: f.reasons,
    }));
  return meta;
}

function redactAnthropicMessages(body, policy) {
  if (!Array.isArray(body?.messages)) return;
  for (const msg of body.messages) {
    if (typeof msg.content === "string" && msg.content) {
      const scan = scanAndRedact(msg.content, policy, msg.role || "user");
      if (scan.status === "flag") msg.content = scan.redacted;
    }
    if (Array.isArray(msg.content)) {
      for (const part of msg.content) {
        if (part?.type === "text" && typeof part.text === "string") {
          const scan = scanAndRedact(part.text, policy, msg.role || "user");
          if (scan.status === "flag") part.text = scan.redacted;
        }
      }
    }
  }
}

export class FirewallAnthropic {
  constructor(client, opts) {
    if (!opts?.policyPath) {
      throw new Error("FirewallAnthropic requires opts.policyPath");
    }
    this._client = client;
    this._policy = loadPolicy(opts.policyPath);
    this.messages = {
      create: async (body, requestOpts) => {
        redactAnthropicMessages(body, this._policy);
        const response = await this._client.messages.create(body, requestOpts);
        const content = response.content || [];
        const calls = content
          .filter((part) => part?.type === "tool_use")
          .map((part) => ({
            id: part.id,
            name: part.name,
            arguments: part.input,
          }));

        response.firewall = enforceToolCalls(calls, this._policy, (blockedIds) => {
          response.content = content.filter(
            (part) => part?.type !== "tool_use" || !blockedIds.has(part.id)
          );
          if (response.content.length === 0) {
            response.content = [{ type: "text", text: this._policy.block_message }];
          }
        });
        return response;
      },
    };
  }

  get models() {
    return this._client.models;
  }
}

export class FirewallGeminiModel {
  constructor(model, policy) {
    this._model = model;
    this._policy = policy;
  }

  async generateContent(request, requestOpts) {
    const response = await this._model.generateContent(request, requestOpts);
    const candidates = read(response, ["response", "candidates"]) || response.candidates || [];
    const calls = [];
    for (const [candidateIndex, candidate] of candidates.entries()) {
      const parts = candidate?.content?.parts || [];
      for (const [partIndex, part] of parts.entries()) {
        const fn = part?.functionCall || part?.function_call;
        if (fn) {
          calls.push({
            id: `gemini-${candidateIndex}-${partIndex}`,
            candidateIndex,
            partIndex,
            name: fn.name,
            arguments: fn.args || fn.arguments || {},
          });
        }
      }
    }

    response.firewall = enforceToolCalls(calls, this._policy, (blockedIds) => {
      const blockedCalls = calls
        .filter((call) => blockedIds.has(call.id))
        .sort((a, b) => b.partIndex - a.partIndex);
      for (const call of blockedCalls) {
        const parts = candidates[call.candidateIndex]?.content?.parts || [];
        parts.splice(call.partIndex, 1);
      }
      for (const candidate of candidates) {
        const parts = candidate?.content?.parts;
        if (Array.isArray(parts) && parts.length === 0) {
          parts.push({ text: this._policy.block_message });
        }
      }
    });
    return response;
  }

  startChat(params = {}) {
    const chat = this._model.startChat(params);
    return new FirewallGeminiChat(chat, this._policy);
  }
}

class FirewallGeminiChat {
  constructor(chat, policy) {
    this._chat = chat;
    this._policy = policy;
  }

  async sendMessage(message, requestOpts) {
    const response = await this._chat.sendMessage(message, requestOpts);
    return new FirewallGeminiModel(
      { generateContent: async () => response },
      this._policy
    ).generateContent({}, requestOpts);
  }
}

export class FirewallGoogleGenerativeAI {
  constructor(client, opts) {
    if (!opts?.policyPath) {
      throw new Error("FirewallGoogleGenerativeAI requires opts.policyPath");
    }
    this._client = client;
    this._policy = loadPolicy(opts.policyPath);
  }

  getGenerativeModel(params, requestOpts) {
    const model = this._client.getGenerativeModel(params, requestOpts);
    return new FirewallGeminiModel(model, this._policy);
  }
}

File: voidhack_agent_defender/providers.py
import json
from typing import Any, Callable

from . import pii
from .client import FirewallOpenAI
from .policy import Policy, load_policy
from .rules import check_tool_calls
from .schemas import Status, ToolCall

OPENAI_COMPATIBLE_BASE_URLS: dict[str, str] = {
    "openai": "https://api.openai.com/v1",
    "groq": "https://api.groq.com/openai/v1",
    "nvidia": "https://integrate.api.nvidia.com/v1",
    "mistral": "https://api.mistral.ai/v1",
    "together": "https://api.together.xyz/v1",
    "fireworks": "https://api.fireworks.ai/inference/v1",
    "perplexity": "https://api.perplexity.ai",
    "deepseek": "https://api.deepseek.com",
    "openrouter": "https://openrouter.ai/api/v1",
    "local": "http://localhost:8000/v1",
}


def create_openai_compatible_firewall(
    provider: str,
    *,
    api_key: str | None = None,
    policy_path: str,
    base_url: str | None = None,
    **client_kwargs: Any,
) -> FirewallOpenAI:
    """Create a guarded OpenAI-compatible client for Groq/NVIDIA/Mistral/Together/etc."""
    resolved_base_url = base_url or OPENAI_COMPATIBLE_BASE_URLS.get(provider)
    if not resolved_base_url:
        raise ValueError(f"Unknown OpenAI-compatible provider: {provider}")
    try:
        from openai import OpenAI
    except ImportError as exc:
        raise ImportError(
            "create_openai_compatible_firewall requires the OpenAI SDK. "
            "Install with `pip install voidhack-agent-firewall[openai]`."
        ) from exc
    raw = OpenAI(api_key=api_key, base_url=resolved_base_url, **client_kwargs)
    return FirewallOpenAI(raw, policy_path=policy_path)


def _stringify_args(value: Any) -> str:
    if isinstance(value, str):
        return value
    try:
        return json.dumps(value if value is not None else {})
    except TypeError:
        return str(value)


def _get(obj: Any, name: str, default: Any = None) -> Any:
    if isinstance(obj, dict):
        return obj.get(name, default)
    return getattr(obj, name, default)


def _set(obj: Any, name: str, value: Any) -> None:
    if isinstance(obj, dict):
        obj[name] = value
    else:
        setattr(obj, name, value)


def _firewall_meta() -> dict[str, Any]:
    return {
        "action": "allow",
        "reason": None,
        "rule_fired": None,
        "stripped_tool_calls": [],
        "blocked_calls": [],
    }


def _enforce_tool_calls(
    calls: list[dict[str, Any]],
    policy: Policy,
    remove_blocked: Callable[[set[str]], None],
) -> dict[str, Any]:
    meta = _firewall_meta()
    if not calls:
        return meta

    tool_calls = [
        ToolCall(
            id=call["id"],
            type="function",
            function={"name": call["name"], "arguments": _stringify_args(call.get("arguments"))},
        )
        for call in calls
    ]
    findings, _ = check_tool_calls(tool_calls, policy)
    blocked_ids = {f.tool_call_id for f in findings if f.status is Status.BLOCK}
    if not blocked_ids:
        return meta

    remove_blocked(blocked_ids)
    first_blocked = next(f for f in findings if f.status is Status.BLOCK)
    meta["action"] = "block"
    meta["stripped_tool_calls"] = sorted(blocked_ids)
    meta["rule_fired"] = "deterministic_rules"
    meta["reason"] = "; ".join(first_blocked.reasons)
    meta["blocked_calls"] = [
        {"name": f.tool_name, "reasons": f.reasons}
        for f in findings
        if f.status is Status.BLOCK
    ]
    return meta


def _redact_anthropic_messages(body: dict[str, Any], policy: Policy) -> None:
    for msg in body.get("messages", []) or []:
        content = msg.get("content")
        role = msg.get("role", "user")
        if isinstance(content, str) and content:
            redacted, check = pii.scan_and_redact(content, policy, source=role)
            if check.status is Status.FLAG:
                msg["content"] = redacted
        elif isinstance(content, list):
            for part in content:
                if isinstance(part, dict) and part.get("type") == "text":
                    text = part.get("text")
                    if isinstance(text, str) and text:
                        redacted, check = pii.scan_and_redact(text, policy, source=role)
                        if check.status is Status.FLAG:
                            part["text"] = redacted


class _AnthropicMessagesWrapper:
    def __init__(self, messages: Any, policy: Policy):
        self._messages = messages
        self._policy = policy

    def create(self, *args: Any, **kwargs: Any) -> Any:
        body = kwargs if kwargs else (args[0] if args and isinstance(args[0], dict) else {})
        _redact_anthropic_messages(body, self._policy)
        response = self._messages.create(*args, **kwargs)
        content = _get(response, "content", []) or []
        calls = [
            {"id": _get(part, "id"), "name": _get(part, "name"), "arguments": _get(part, "input")}
            for part in content
            if _get(part, "type") == "tool_use"
        ]

        def remove(blocked_ids: set[str]) -> None:
            kept = [
                part
                for part in content
                if _get(part, "type") != "tool_use" or _get(part, "id") not in blocked_ids
            ]
            if not kept:
                kept = [{"type": "text", "text": self._policy.block_message}]
            _set(response, "content", kept)

        _set(response, "firewall", _enforce_tool_calls(calls, self._policy, remove))
        return response

    def __getattr__(self, name: str) -> Any:
        return getattr(self._messages, name)


class FirewallAnthropic:
    """Drop-in wrapper for Anthropic Claude clients using native tool_use blocks."""

    def __init__(self, client: Any, policy_path: str):
        self._client = client
        self._policy = load_policy(policy_path)
        self.messages = _AnthropicMessagesWrapper(client.messages, self._policy)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._client, name)


class FirewallGeminiModel:
    """Wrapper for Gemini model objects that strips blocked function_call parts."""

    def __init__(self, model: Any, policy: Policy):
        self._model = model
        self._policy = policy

    def generate_content(self, *args: Any, **kwargs: Any) -> Any:
        response = self._model.generate_content(*args, **kwargs)
        candidates = _get(response, "candidates", []) or []
        calls: list[dict[str, Any]] = []
        for candidate_index, candidate in enumerate(candidates):
            content = _get(candidate, "content")
            parts = _get(content, "parts", []) if content is not None else []
            for part_index, part in enumerate(parts):
                fn = _get(part, "function_call") or _get(part, "functionCall")
                if fn is not None:
                    calls.append(
                        {
                            "id": f"gemini-{candidate_index}-{part_index}",
                            "candidate_index": candidate_index,
                            "part_index": part_index,
                            "name": _get(fn, "name"),
                            "arguments": _get(fn, "args") or _get(fn, "arguments") or {},
                        }
                    )

        def remove(blocked_ids: set[str]) -> None:
            for call in sorted(calls, key=lambda c: c["part_index"], reverse=True):
                if call["id"] not in blocked_ids:
                    continue
                candidate = candidates[call["candidate_index"]]
                parts = _get(_get(candidate, "content"), "parts", [])
                if isinstance(parts, list):
                    parts.pop(call["part_index"])
                    if not parts:
                        parts.append({"text": self._policy.block_message})

        _set(response, "firewall", _enforce_tool_calls(calls, self._policy, remove))
        return response

    def start_chat(self, *args: Any, **kwargs: Any) -> Any:
        chat = self._model.start_chat(*args, **kwargs)
        return FirewallGeminiChat(chat, self._policy)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._model, name)


class FirewallGeminiChat:
    def __init__(self, chat: Any, policy: Policy):
        self._chat = chat
        self._policy = policy

    def send_message(self, *args: Any, **kwargs: Any) -> Any:
        response = self._chat.send_message(*args, **kwargs)
        return FirewallGeminiModel(
            type("_OneShotGeminiModel", (), {"generate_content": lambda *_a, **_k: response})(),
            self._policy,
        ).generate_content()

    def __getattr__(self, name: str) -> Any:
        return getattr(self._chat, name)


class FirewallGoogleGenerativeAI:
    """Drop-in wrapper for google-generativeai clients."""

    def __init__(self, client: Any, policy_path: str):
        self._client = client
        self._policy = load_policy(policy_path)

    def GenerativeModel(self, *args: Any, **kwargs: Any) -> FirewallGeminiModel:
        return FirewallGeminiModel(self._client.GenerativeModel(*args, **kwargs), self._policy)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._client, name)
